flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/CheckpointBarrierHandler.java

```diff
@@ -153,6 +153,8 @@ protected void markAlignmentEnd() {
     }
 
     protected void markAlignmentEnd(long alignmentDuration) {
+        checkState(alignmentDuration >= 0 || latestAlignmentDurationNanos.isDone());
```

Comment:
What is the case when `alignmentDuration < 0 && !latestAlignmentDurationNanos.isDone()` ? Is it a valid scenario? If it is could you add a comment for it here?

Lastly, would be good to give some kind of a message here in case it is hit by a user.

Reply:
In my opinion, it is only possible when the time is not monotonic. In the current implementation, if this happens we will fail here (org/apache/flink/runtime/checkpoint/CheckpointMetrics.java:74) - which is too late and too difficult to debug. as I understand, we have two main options here:

- Adding the check condition what I've done so we fail if we receive a negative value.
- Removing the existing check from CheckpointMetrics.java:74(we can only log the warning), so in this case, if the time is not monotonic(or we have another bug) we won't fail but just display the negative time. We should discuss it because perhaps it is a better solution than mine.

Reply:
Ok, I get the negative value for `alignmentDuration`, but what is the role of `latestAlignmentDurationNanos` here?

Reply:
Assuming that `alignmentDuration < 0`, why does it matter if `latestAlignmentDurationNanos.isDone()` or `!latestAlignmentDurationNanos.isDone()`?  Shouldn't it be just `checkState(!latestAlignmentDurationNanos.isDone());`?

Reply:
You can take a look at this test org.apache.flink.streaming.runtime.io.checkpointing.CheckpointBarrierTrackerTest#testCompleteCheckpointsOnLateBarriers, especially at the last barriers of checkpoint 3 and 4. According to this test, it is possible to have two last barriers one by one which makes possible the situation when `alignmentDuration < 0`(default value) but it can happen only when `latestAlignmentDurationNanos` is already done.
In fact, I think that CheckpointBarrierTracker implemented incorrectly because it supports several checkpoints in progress but `markAlignmentStart` and `markAlignmentEnd` support only one checkpoint at once. I don't sure that it makes sense to fix this behavior in this ticket but in general, if you agree that it is incorrect we can create a separate ticket for this.

Reply:
> You can take a look at this test CheckpointBarrierTrackerTest#testCompleteCheckpointsOnLateBarriers, especially at the last barriers of checkpoint 3 and 4. According to this test, it is possible to have two last barriers one by one

I think indeed the alignment duration in this case is not calculated perfectly. For `CheckpointBarrierTracker` ideally we should calculate the alignment duration independently for each of the checkpoints, but it doesn't sound like a worthwhile change to me. Anyway, in that case, when we have checkpoints 3 and 4 closely following one another, `alignmentDuration < 0` still shouldn't happen. Either for the alignmentDuration we should track only the oldest or the most recent checkpoint.

As I understand currently we can first call `markAlignmentStart` twice in a row for chk 3 and chk 4, and after that `markAlignmentEnd` can be called twice also in a row for both chk 3 and chk 4? And this second `markAlignmentEnd` hits this condition that the `!latestAlignmentDurationNanos.isDone()` and `alignmentDuration` can overflow?

Reply:
> Either for the alignmentDuration we should track only the oldest or the most recent checkpoint.

Yes, I guess we can change the `CheckpointBarrierTracker` in such a way to ignore the `markAlignmentStart' and `markAlignmentEnd` if it relates to the checkpoint which less than highest one.

> As I understand currently we can first call markAlignmentStart twice in a row for chk 3 and chk 4, and after that markAlignmentEnd can be called twice also in a row for both chk 3 and chk 4? And this second markAlignmentEnd hits this condition that the !latestAlignmentDurationNanos.isDone() and alignmentDuration can overflow?

Yes, it is correct